reporter/attrmgr.go

```diff
@@ -0,0 +1,62 @@
+package reporter
+
+import (
+	"fmt"
+
+	common "go.opentelemetry.io/proto/otlp/common/v1"
+	profiles "go.opentelemetry.io/proto/otlp/profiles/v1experimental"
+)
+
+// AttrIndex is an index in the `Profile.attribute_table`.
+type AttrIndex = uint64
+
+// AttrTableManager maintains the `Profile.attribute_table` field for the
+// `profile.Profile` that it was created for. Attributes are automatically
+// deduplicated by the manager.
+type AttrTableManager struct {
+	// indices maps compound keys to the indices in the attribute table.
+	indices map[string]AttrIndex
+
+	// profile being built right now.
+	profile *profiles.Profile
+}
+
+func NewAttrTableManager(profile *profiles.Profile) *AttrTableManager {
+	return &AttrTableManager{
+		indices: make(map[string]AttrIndex),
+		profile: profile,
+	}
+}
+
+func (m *AttrTableManager) AddIntAttr(key string, value int64) AttrIndex {
+	compound := fmt.Sprintf("%v_%d", key, value)
+	val := common.AnyValue{Value: &common.AnyValue_IntValue{IntValue: value}}
+	return m.addAnyAttr(key, compound, &val)
+}
+
+func (m *AttrTableManager) AddStringAttr(key, value string) AttrIndex {
+	compound := fmt.Sprintf("%v_%s", key, value)
+	val := common.AnyValue{Value: &common.AnyValue_StringValue{StringValue: value}}
+	return m.addAnyAttr(key, compound, &val)
+}
+
+func (m *AttrTableManager) addAnyAttr(
+	key string,
+	compoundKey string,
```

Comment:
Simplification: You can build the compoundKey inside the function with
```
	compoundKey := fmt.Sprintf("%v_%v", key, value)
```

Reply:
I originally considered that, but there's a bunch of private protobuf fields in `AnyValue` that will make the key unnecessarily long. It also ends up printing pointers, so it is prone to break if Go were to switch to a moving GC. Example formatting of `AnyValue` object:

```
{{{} [] [] <nil>} 0 [] 0xc00011f610}{{{} [] [] <nil>} 0 [] 0xc00011f630}{{{} [] [] <nil>} 0 [] 0xc00011f660}
```

Reply:
When using value.GetValue() it looks like
```
service.name_&{apmServiceName2}
process.pid_&{6789}
```
Not sure why this should change with a moving GC.

Reply:
Ah yeah, with `GetValue` it looks differently, since `AnyValue_StringValue` and `AnyValue_IntValue` don't have the private protobuf meta fields. Looks like they aren't generated for the `oneof` field structs (probably reuse cache in parent).

I can add what you suggest, but honestly I don't think it makes the code any better, since it relies on a lot of implicit assumptions. If we were to add support for `AnyValue_ArrayValue` (that contains a pointer) later, it'd be unsafe again because it will compare pointers instead of contents, with even the same instance not being guaranteed to be considered equal if a moving GC was introduced.

So to save a grand total of 2 lines of code, we'd now have introduced a bunch of possible future bugs.

Reply:
Not sure how likely `AnyValue_ArrayValue` becomes added here, but feel free to stay with the current code. It has the advantage that we the option to optimize building the compound string when Sprintf turns out to be a slow horse (variadic arguments always escape to the heap last time I tested).